### ISSUE TITLE
Improve Dependency Resolution for Workflow Optimization

### ISSUE THREAD

**Issue by trajan0x** [synapsecns/sanguine#2235]
### Title:
Improve Dependency Resolution for Workflow Optimization

#### Summary:
The current implementation of our GitHub Action, as defined in [sanguine/contrib/git-changes-action](https://github.com/synapsecns/sanguine/tree/master/contrib/git-changes-action), uses `go.mod` to determine dependencies for triggering workflows. This broad approach can lead to unnecessary workflow runs, as changes in dependencies that do not directly affect the module in question still trigger the workflow. We propose enhancing our dependency resolution mechanism to be more granular, focusing on package-level dependencies rather than module-level dependencies defined in `go.mod`.

#### Problem:
Currently, if any dependency within a module changes, our workflow is triggered, regardless of whether the change directly impacts the packages our module uses. This leads to inefficient use of resources, as workflows run for changes that are irrelevant to the module's operation. For example, if our module `A` directly depends on package `A` from module `B`, we want to trigger the workflow only if package `A` (or any package within `A`'s dependency tree) changes. Changes in module `B` that do not affect package `A` should not trigger our module `A`'s workflow.

#### Proposed Solution:
Implement a more refined dependency tracking system that:
1. Identifies direct package-level dependencies for each module within our project.
2. Triggers workflows only when there are changes in the directly dependent packages or any package within their dependency tree.
3. Excludes broader module-level changes from triggering workflows when the changes do not affect the directly used packages.

This approach aims to optimize our CI/CD process by reducing unnecessary workflow runs, thereby saving computational resources and developer time.

#### Expected Benefits:
- **Efficiency:** Reduces unnecessary CI/CD runs, focusing only on relevant changes.
- **Resource Optimization:** Better use of computational resources, decreasing overall run times and potential costs.
- **Developer Productivity:** Developers can focus on changes that directly impact their work, reducing noise from unrelated workflow runs.

#### Request:
We seek input from the community and the project's maintainers on the feasibility of this approach and any suggestions for implementation strategies. Additionally, we're looking for contributors interested in developing this enhancement.

**Comment by oliverqx** [comment]
can i get assigned?

**Comment by trajan0x** [comment]
@oliverqx apologies, this is my first time using algora. Would you happen to kno the slash command for that?

**Comment by trajan0x** [comment]
oh I see it.

**Comment by oliverqx** [comment]
thanks!

**Comment by shashank40** [comment]
Hey @oliverqx. All the best for your PR. 
But can I pick it up or you are working on it?

**Comment by oliverqx** [comment]
im currently working on the solution, feel confident about the ETA i previously shared

**Comment by trajan0x** [comment]
love to hear it

**Comment by oliverqx** [comment]
@trajan0x just to give an update,

I think i might need a couple more days, testing my changes on github actions has been a bit more challenging than i predicted but I'm almost there. Will open a PR tomorrow and keep working. I'm pretty sure I'll be done by saturday. I was doing this half time and the next two days i can go full time into it

**Comment by shashank40** [comment]
If you guys are okay i can give it a try as well @oliverqx @trajan0x

**Comment by trajan0x** [comment]
@oliverqx got it. If there's anything I can answer to help lmk

@shashank40 gonna leave that up to @oliverqx

**Comment by shashank40** [comment]
Till then @trajan0x you can point me to other important issues?

**Comment by shashank40** [comment]
Any update @oliverqx ?

**Comment by oliverqx** [comment]
I'm working on it, I'll open my PR in a couple of hours

**Comment by trajan0x** [comment]
@shashank40 have something for you to try out if you're reasonably comfortable w/ docker

**Comment by shashank40** [comment]
Sure @trajan0x. I can pick Docker. I have worked with it quite a few times.

**Comment by trajan0x** [comment]
@shashank40  actually, had a different idea first. Please see #2318

@oliverqx just checking in here

**Comment by oliverqx** [comment]
hey @trajan0x thank you for your patience. I think I'm finally done. Doing this part time was quite a challenge as I kept loosing my train of thought, and that pushed me back a couple of steps each time I did.

I've created this [dummy package](https://github.com/oliverqx/showcase) that showcases the changes I've made to git-changes-action. 

Just a heads up, I thought having some sort of categorical relation between modules and packages would make it easier to follow their dependencies, and only later noticed this naming convention is very awkward. But I think both the readme file and opened PRs explain everything quite well.

Only two things left to do: 
1. clean recursion for deep dependency resolution
2. migrate my changes to synapsecns' sanguine repo, this is because I made a clean copy of the go project to test changes more efficiently. The changes I've made so far can be seen [here](https://github.com/oliverqx/git-changes-action). But I'll be opening a formal PR after getting some sleep.
